File: 백준/Gold/9082. 지뢰찾기/지뢰찾기.py
t = int(input())
INF = 1e9


def main(hide, bomb, first, second):
    cnt = first + second
    for i in range(1, n-1):
        x = hide[i] - first - second
        if x == 0:
            first = second
            second = 0
        elif x == 1:
            first = second
            second = 1
            cnt += 1
        else:
            return 0

    # if hide[n-1] == 0:
    #     if first + second != 0:
    #         return 0
    # elif hide[n-1] == 1:
    #     if first + second != 1:
    #         return 0
    # elif hide[n-1] == 2:
    #     if first + second != 2:
    #         return 0
    return cnt


for _ in range(t):
    n = int(input())
    hide = [int(p) for p in input()]
    bomb = [p for p in input()]

    box = []

    if hide[0] == 0:
        box.append(main(hide, bomb, 0, 0))
    elif hide[0] == 1:
        box.append(main(hide, bomb, 1, 0))
        box.append(main(hide, bomb, 0, 1))
    elif hide[0] == 2:
        box.append(main(hide, bomb, 1, 1))

    print(max(box))

File: 백준/Gold/9082. 지뢰찾기/지뢰찾기.cc
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;
#define FIO                                                                    \
  ios_base::sync_with_stdio(false);                                            \
  cin.tie(NULL);                                                               \
  cout.tie(NULL);
typedef long long ll;

int n, t;
string num, bomb;

int check(string num, string bomb, int cnt) {
  if (cnt)
    bomb[0] = '*';
  else
    bomb[0] = '.';

  int siz = num.size();
  int answer = 0;
  for (int i = 0; i < siz; i++) {
    if (i != 0 && i != siz - 1) {
      int count = 0;
      for (int j = i - 1; j < i + 2; j++) {
        if (bomb[j] == '*')
          num[i]--;
      }
      if (num[i] == '0')
        ;
      else if (num[i] == '1')
        bomb[i + 1] = '*';
      else if (num[i] == '2')
        return 0;
    } else if (i == 0) {
      int count = 0;
      for (int j = i; j < i + 2; j++) {
        if (bomb[j] == '*')
          num[i]--;
      }
      if (num[i] == '0')
        ;
      else if (num[i] == '1')
        bomb[i + 1] = '*';
      else if (num[i] == '2')
        return 0;

    } else if (i == siz - 1) {
      int count = 0;
      for (int j = i - 1; j < i + 1; j++) {
        if (bomb[j] == '*')
          num[i]--;
      }
      if (num[i] == '0')
        ;
      else if (num[i] == '1')
        return 0;
      else if (num[i] == '2')
        return 0;
    }
    if (bomb[i] == '*')
      answer++;
  }
  // for (auto i : bomb)
  //   cout << i;
  return answer;
}
int main() {
  FIO;
  cin >> t;
  while (t--) {
    cin >> n >> num >> bomb;
    int answer = 0;
    answer = max(answer, check(num, bomb, 0));
    if (num[0] - '0' > 0)
      answer = max(answer, check(num, bomb, 1));
    cout << answer << "\n";
  }
}